lib/generator/files.rb

```diff
@@ -0,0 +1,56 @@
+require 'pathname'
+
+module Generator
+  module Files
+    class Readable
+      attr_reader :filename, :repository_root
+      def initialize(filename:,repository_root: nil)
+        @filename = filename
+        @repository_root = repository_root
+      end
+
+      def to_s
+        File.read(filename) if File.exist?(filename)
+      end
+
+      def abbreviated_commit_hash
+        GitCommand.abbreviated_commit_hash(git_path, relative_filename)
+      end
+
+      private
+
+      def relative_filename
+        Pathname.new(filename).relative_path_from(Pathname.new(repository_root)).to_s
+      end
+
+      def git_path
+        File.join(repository_root, '.git')
+      end
+    end
+
+    class Writable < Readable
+      def save_if_changed(new_content)
```

Comment:
The name should be `save`.  We don't care if it is changed or not, but for efficiency, internally, it can make that distinction.  Also, if it is new content, it is not changed.  `content` should be sufficient.  Unless it is truly requiring that it is indeed new content.